File: completed-labs/06/java/Queue.java
public class Queue {
    private SLList head = SLList.NIL;
    private SLList tail = SLList.NIL;
    public Queue() { }
    public void enqueue(Object o) {
	if (tail == SLList.NIL) {
	    tail = new SLList(o, tail);
	    head = tail;
	} else {
	    tail.setRest(new SLList(o, SLList.NIL));
	    tail = tail.rest();
	}
    }
    public Object dequeue() {
	Object r = head.first();
	head = head.rest();
	if (head == SLList.NIL) {
	    tail = SLList.NIL;
	}
	return r;
    }
    public Object head() {
	return head.first();
    }
    public boolean empty() {
	return head == SLList.NIL;
    }
}

File: is52038b-labs/06/cpp/Queue.cpp
#include "SLList.hpp"
#include "Queue.hpp"

using namespace std;

void Queue::enqueue(int o) {
  if (tail == SLList::NIL) {
    tail = new SLList(o, tail);
    hd = tail;
  } else {
    tail->setRest(new SLList(o, SLList::NIL));
    tail = tail->rest();
  }
}

int Queue::dequeue() {
  int r = hd->first();
  SLList *newhd = hd->rest();
  delete hd;
  hd = newhd;
  if (hd == SLList::NIL) {
    tail = SLList::NIL;
  }
  return r;
}

int Queue::head() {
  return hd->first();
}

bool Queue::empty() {
  return hd == SLList::NIL;
}
